### PR TITLE
Build wheels automatically in CI environment

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Right now when it's time to push a relase of qiskit-terra we have to
manually build wheels for all our supported environments because we
included compiled code in the project. This involves setting up a proper
environment with portable librariers and compilers in Linux, Windows,
and Mac OSX with all 3 versions of python we support. Then manually
building the 9 wheels necessary to publish on pypi. However we install
qiskit-terra from source every day multiple times in all these
environments in our CI. There is no reason this manual step needs to
exist and we can just leverage CI to do our wheel building when it's
time to do a release.

This commit adds initial support for building our wheels in travis
(for linux and osx) and appveyor (for windows). It leverages the
cibuildwheel project to automate the wheel building. This will ensure we
setup the proper environment for all 3 python versions on all 3
environments (for example pulling in the manylinux docker container on
linux).

Building off this in the future it should only be necessary to push a tag on
release and the automation will take care of building all the release artifacts
and pushing them to pypi.

### Details and comments
Things to do still:

- [x]  Find a place to upload to for testing travis wheels locally (appveyor stores
    artifacts) for verification (done by using cibw to just run tests on the built wheel)
- [x] Setup automation around tag pushes to avoid building wheels on every
    PR and/or push event.

### CODE REVIEW

**Comment by mtreinish** [comment]
Ok, 2/3 of the builds are working now. Mac OSX is still failing because it's ignoring the skip env var and trying to build python 2.7. 

Since appveyor allows artifact uploads you can see download the results of those wheel builds here: https://ci.appveyor.com/project/Qiskit/qiskit-sdk-py/builds/22931221/job/j3fllycet9nvo4k1/artifacts

After fixing OS X the next step is to find a place to upload the travis builds so we can locally test the output from everything to verify this does the right thing.

**Comment by atilag** [comment]
This is great, Aer comes next :)

**Comment by mtreinish** [comment]
Ok, all the wheel builds are completing successfully now. The next test is to setup some external cloud storage for travis to upload the output to so we can verify the output is as we expect. Once that manual verification is complete we can change the job config to only run the wheel builds on the stable branch and on tag pushes upload to pypi.

**Comment by mtreinish** [comment]
Still need to add the appveyor logic bits to make it only run wheel builds on tags, but this is almost ready

**Comment by mtreinish** [comment]
I think the appveyor side is correct, but it won't be triggered except for on a tag so I won't know until release.

**Comment by mtreinish** [comment]
Oh, I'm being silly. I can just remove the tag conditional to test it. I'll switch this back to WIP test that and only remove the WIP until it works as expected.

**Comment by atilag** [comment]
Great job Matthew, I'm reviewing this because I want to implement something similar in Aer.

**Comment by mtreinish** [comment]
Ok, this should be ready to go I think. I can't test the twine upload to test.pypi.org bits until we actually push a tag unfortunately (we'd be able to if we had dev versions, as discussed in #1796 ). But at least the wheel builds look as though they're working as expected.

The workflow for release with this should be tag 0.8 (and push to github). Then we should be able to download the wheels from https://test.pypi.org/project/qiskit-terra/#files after the builds finishm inspect/test them and then push them with an sdist to the real pypi.

File: .travis.yml [hunk 1]
```diff
@@ -83,6 +83,35 @@ stages:
...
 jobs:
   include:
+    # Wheel builds
+    - sudo: required
+      services:
+        - docker
+      env:
+        - CIBW_BEFORE_BUILD="pip install -U Cython"
+        - CIBW_SKIP="cp27-* cp34-*"
+        - TWINE_REPOSITORY_URL=https://test.pypi.org/legacy/
```
Comment (atilag) [inline .travis.yml]: Do you think that we should be pointing to the test pypi servers and try this for a while to make sure that everything works as expected? ... or you are very confident about this :)

Comment (mtreinish) [reply .travis.yml]: My thinking was that we'd use test.pypi.org as mostly a data store so we can download and test the artifacts before publishing a release to the production pypi. While I'm sure the full automatic approach would work fine, having the manual step in the middle is nice to just make sure it all worked as expected.

**Pull request merged**